src/ucs/datastruct/ptr_array.h

```diff
@@ -121,12 +159,236 @@ void *ucs_ptr_array_replace(ucs_ptr_array_t *ptr_array, unsigned index, void *ne
                     !__ucs_ptr_array_is_free(_var = (void*)((_ptr_array)->start[_index])))
 
 
+/**
+ * For-each user function: Calculates how many free elements are ahead.
+ *
+ * @param [in] ptr_array      Pointer to a ptr array.
+ * @param [in] element_index  Index of slot
+ *
+ * @return size_elem - The number of free elements ahead if free, if not 1.
+ */
+static inline uint32_t
+__ucs_ptr_array_for_each_get_step_size(ucs_ptr_array_t *ptr_array,
+                                       unsigned element_index)
+{
+    uint32_t size_elem;
+    ucs_ptr_array_elem_t elem = ptr_array->start[element_index];
+
+    if (ucs_unlikely(__ucs_ptr_array_is_free((ucs_ptr_array_elem_t)elem))) {
+       size_elem = (elem >> UCS_PTR_ARRAY_SIZE_FREE_SHIFT);
+    } else {
+       size_elem = 1;
+    }
+    ucs_assert(size_elem > 0);
+
+    return size_elem;
+}
+
+
+/**
+ * Check if element is free.
+ *
+ * @param [in] _elem        An element in the ptr array.
+ *
+ * @return 1 if the element is free and 0 if it's occupied.
+ */
+#define __ucs_ptr_array_is_free(_elem) ((uintptr_t)(_elem) & UCS_PTR_ARRAY_FLAG_FREE)
+
+
 /**
  * Iterate over all valid elements in the array.
+ *
+ * @param [out] _var        Pointer to current array element in the foreach.
+ * @param [out] _index      Index variable to use as iterator (unsigned).
+ * @param [in]  _ptr_array  Pointer to a ptr array.
  */
 #define ucs_ptr_array_for_each(_var, _index, _ptr_array) \
-    for (_index = 0; _index < (_ptr_array)->size; ++_index) \
-         if (!__ucs_ptr_array_is_free(_var = (void*)((_ptr_array)->start[_index]))) \
+    for ((_index) = 0; ((_index) < (_ptr_array)->size); \
+         (_index) += __ucs_ptr_array_for_each_get_step_size((_ptr_array), (_index))) \
+         if ((ucs_likely(!__ucs_ptr_array_is_free( \
+             (ucs_ptr_array_elem_t)((_var) = (void *)((_ptr_array)->start[(_index)]))))))
```

Comment:
would be better to assign _var during iteration step and not during condition
also, does it assume index 0 is never a "hole"?